assert: .throws accept objects

BridgeAR opened the pull request:
From now on it is possible to use a validation object in throws
instead of the other possibilities.

Refs #17557 

I personally am not sure if this is a cool thing or not but I thought I just put something together to get some votes.

CI https://ci.nodejs.org/job/node-test-pull-request/12016/

<!--
Thank you for your pull request. Please provide a description above and review
the requirements below.

Bug fixes and new features should include tests and possibly benchmarks.

Contributors guide: https://github.com/nodejs/node/blob/master/CONTRIBUTING.md
-->

##### Checklist
<!-- Remove items that do not apply. For completed items, change [ ] to [x]. -->

- [x] `make -j4 test` (UNIX), or `vcbuild test` (Windows) passes
- [x] tests and/or benchmarks are included
- [x] documentation is changed or added
- [x] commit message follows [commit guidelines](https://github.com/nodejs/node/blob/master/CONTRIBUTING.md#commit-message-guidelines)

##### Affected core subsystem(s)
<!-- Provide affected core subsystem(s) (like doc, cluster, crypto, etc). -->
assert

Trott commented on doc/api/assert.md:
```diff
@@ -711,18 +711,21 @@ instead of the `AssertionError`.
-If specified, `error` can be a constructor, [`RegExp`][], or validation
-function.
+If specified, `error` can be a constructor, [`RegExp`][], a validation
+function, or a object where each property will be tested for.
```
Nit: `a object` -> `an object`

Trott commented:
Thanks for doing this. I like it. This should make it possible to simplify (or even eliminate?) `common.expectsError()`.

BridgeAR commented:
I addressed the comment and improved some things further. Right now I still do not see how it will replace or simplify `common.expectsError` though.

The differences as I see them:
`common.expectsError`:
1. accepts special input e.g. a RegExp for _some_ fields (e.g. `message`).
1. has a mandatory check on the `code` property.
1. can be used as callback function that also works like `common.mustCall`.
1. strictly verifies that the `type` property, if present, is also inherited by `Error` (or is `Error`).
1. verifies that the error is a instance of `type` (if `type` is present).
1. only checks a limited amount of fields no matter if there are more on the provided object.

So we could overcome some of these e.g.
- (1) - We could special handle either especially the `message` field to also accept a `RegExp` or to always accept a `RegExp` in case the field to test contains a `string`.
- (3) - by continuing to use `common.expectsError` in exactly that case or we could add that functionality to `assert.throws` as well.
- (6) - is actually better here out of my perspective.

(4) and (5) are probably not really that important even though the replacement would only be to check for the name, but that would still be good enough out of my perspective. I do not see any solution for (2) though.

Trott commented:
I'm not sure I like the mandatory `code` check, to be honest. It means that `common.expectsError()` can only be used on internal Node.js errors and not on errors from the JS engine.  But this can be a later discussion. Even if all it does is make it easier to switch between `assert.throws()` and `common.expectsError()`, that's still a big enough win for me. :-D

BridgeAR commented:
Any further votes on if this is a good or bad idea? @nodejs/collaborators PTAL

ronkorving commented on lib/assert.js:
```diff
@@ -136,10 +137,44 @@ assert.notStrictEqual = function notStrictEqual(actual, expected, message) {
+        createMsg(msg, 'message', actual, expected));
+    }
+    const keys = Object.keys(expected);
+    for (const key of keys) {
```
`name` and `message` have already been tested for, so perhaps they should be explicitly not tested for again? I imagine that testing for them would be a very common scenario, in which case not running `deepStrictEqual` will probably run a bit faster. Thoughts?

BridgeAR replied:
I guess the performance will be exactly the same. `deepStrictEqual` will only do

```js
// Already existing
function equal(a, b) { if (a === b) return true }

// This check is executed
if (equal(a, b) !== true) throw new Error(msg)
```

BridgeAR commented:
Rebased. New CI https://ci.nodejs.org/job/node-test-commit-light/43/
(Only the small one to verify the rebase as the full run did not show anything to look further into).

BridgeAR commented:
CI again https://ci.nodejs.org/job/node-test-commit-light/48/

BridgeAR commented:
Landed in 2d374916ebbaaa83aaf68577b75d41c6bb6ca9b8

MylesBorins commented:
This does not land cleanly on v9.x, would someone be willing to backport?

BridgeAR commented:
Backport opened in #19230

MylesBorins commented:
Requested backport to 8.x in #19230
